stoobly_agent/app/cli/scaffold/service_workflow_validate_command.py

```diff
@@ -0,0 +1,226 @@
+from stoobly_agent.config.data_dir import DATA_DIR_NAME
+
+from .app import App
+
+
+class ServiceWorkflowValidateCommand(ServiceCommand, ValidateCommand):
+  def __init__(self, app: App, **kwargs):
+    ServiceCommand.__init__(self, app, **kwargs)
+    ValidateCommand.__init__(self)
+
+    self.workflow_name = kwargs['workflow_name']
+    self.hostname = self.service_config.hostname
+    self.service_docker_compose = ServiceDockerCompose(app_dir_path=app.dir_path, target_workflow_name=self.workflow_name, service_name=self.service_name, hostname=self.hostname)
+
+  @property
+  def fixtures_dir_path(self):
+    return os.path.join(self.workflow_path, FIXTURES_FOLDER_NAME)
+  
+  @property
+  def workflow_path(self):
+    return os.path.join(
+      self.scaffold_dir_path,
+      self.workflow_relative_path
+    )
+  @property
+  def workflow_relative_path(self):
+    return os.path.join(
+      self.service_relative_path,
+      self.workflow_name
+    )
+
+  def is_local(self):
+    with open (self.service_docker_compose.docker_compose_path,'rb') as f:
+      docker_compose_file_content = yaml.safe_load(f)
+      if docker_compose_file_content and docker_compose_file_content.get('services'):
+        return True
+
+      # We can potentially check the port too someday
+
+    return False
+
+  def is_external(self):
+    return not self.is_local()
+
+  def hostname_reachable(self, url: str) -> None:
+    # Retry HTTP request. Source: https://stackoverflow.com/questions/15431044/can-i-set-max-retries-for-requests-request
+    s = requests.Session()
+    retries = Retry(total=5,
+                    backoff_factor=0.1,
+                    status_forcelist=[ 500, 502, 503, 504 ])
+    s.mount('http://', HTTPAdapter(max_retries=retries))
+    response = s.get(url=url)
+
+    if not response.ok:
+      raise ScaffoldValidateException(f"Host is not reachable: {url}")
+ 
+  def validate_hostname(self, url: str) -> None:
+    print(f"Validating hostname: {url}")
+    self.hostname_reachable(url)
+
+    # TODO: check logs of proxy. lifecycle hook for custom logging? Does mitmproxy support json logging?
+
+  def validate_internal_hostname(self, url: str) -> None:
+    print(f"Validating hostname inside Docker network, url: {url}")
+    
+    timeout_seconds = 1
+    output = self.docker_client.containers.run(
+      image='curlimages/curl:8.11.0',
+      command=f"curl --max-time {timeout_seconds} {url} --verbose",
+      network=self.app_config.network,
+      stderr=True,
+      remove=True,
+    )
+
+    # Note: 499 error could also mean success because it shows the proxy
+    # connection is working, but we haven't recorded anything yet
+    logs = output.decode('ascii')
+    if ('200 OK' not in logs) and ('499' not in logs):
+      raise ScaffoldValidateException(f"Error reaching {url} from inside Docker network")
+
+  # TODO: might be better in WorkflowValidateCommand
```

Comment:
Are we differentiating `ServiceWorkflowValidateCommand` and `WorkflowValidateCommand` by whether or not `ServiceCommand` is referenced? If so, then agree, this may be better in the latter since it doesn't seem to reference `ServiceCommand`.

Reply:
The difference is `WorkflowValidateCommand` is for validating for everything not related to services such as core components like the gateway. `ServiceWorkflowValidateCommand` is for services

Right now there the proxy container validation is in `ServiceWorkflowValidateCommand` which we need to know the service name to know the proxy container name